File: 2016/src/11.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::collections::VecDeque;
use std::fs::File;
use std::io::Read;

#[derive(Clone, Debug, PartialEq)]
enum Thing {
    Generator(String),
    Microchip(String),
}

fn parse_line(text: &str) -> (u32, Vec<Thing>) {
    let mut floor = 0u32;
    let mut adjective: String = String::new();
    let mut things: Vec<Thing> = vec![];
    
    // remove punctuation
    let mut line = String::from(text);
    line.retain(|c| c != '.' && c != ',');

    for part in line.split(" ") {
        match part {
            // Floor numbers
            "first" => floor = 1,
            "second" => floor = 2,
            "third" => floor = 3,
            "fourth" => floor = 4,
            // Uninteresting sauce
            "The" | "floor" | "a" | "contains" | "nothing" | "relevant" | "and" => (),
            // Keywords
            "generator" => {
                assert!(!adjective.is_empty());
                things.push(Thing::Generator(adjective));
                adjective = String::new();
            },
            "microchip" => {
                assert!(!adjective.is_empty());
                // Remove the "-compatible" suffix
                // TODO: remove_matches - as of rustc 1.53.0, it's an experimental feature
                //adjective.remove_matches("-compatible");
                let suffix = adjective.find("-compatible").unwrap();
                adjective.truncate(suffix);
                things.push(Thing::Microchip(adjective));
                adjective = String::new();
            },
            // Adjectives
            &_ => adjective = part.to_string(),
        };
    }
    (floor, things)
}

fn parse_input(text: &str) -> HashMap<u32, Vec<Thing>> {
    let mut floors: HashMap<u32, Vec<Thing>> = HashMap::new();
    for line in text.lines() {
        let (num, contents) = parse_line(&line);
        println!("Floor {}, contents {:?}", num, contents);
        assert!(num >= 1 && num <= 4);
        assert!(!floors.contains_key(&num));
        floors.insert(num, contents);
    }
    floors
}

#[derive(Clone, Default)]
struct State {
    floors: [Vec<Thing>; 4],
    current: usize,
}

fn make_state(floor_map: HashMap<u32, Vec<Thing>>) -> State {
    let mut state: State = Default::default();
    floor_map.into_iter().for_each(|(k, v)| -> () {
        assert!(k >= 1 && k <= 4);
        state.floors[(k-1) as usize] = v;
    });
    state
}

// Shorten Thing to just 'G' (for generator) or 'M' (for microchip)
// and the kind.
fn print_thing(thing: &Thing) -> String {
    match thing {
        Thing::Generator(x) => "G".to_string() + x,
        Thing::Microchip(x) => "M".to_string() + x,
    }
}

impl State {
    // Serialize the State into a String. The strings can be compared
    // to check if two states are equal
    pub fn as_string(&self) -> String {
        self.floors.iter()
                   .map(|floor| {
                        let mut items = floor.iter()
                             .map(|thing| print_thing(thing))
                             .collect::<Vec<_>>();
                        items.sort();
                        items.join(",")
                   })
                   .collect::<Vec<_>>()
                   .join("::")
            + " " + &self.current.to_string()
    }

    pub fn is_final(&self) -> bool {
        self.floors[0].len() == 0
        && self.floors[1].len() == 0
        && self.floors[2].len() == 0
    }
}

fn pick_one_or_two(items: &Vec<Thing>) -> Vec<(Vec<Thing>, Vec<Thing>)> {
    let mut result: Vec<(Vec<Thing>, Vec<Thing>)> = Vec::new();
    for i in 0 .. items.len() {
        // Pick 1 item
        result.push((
            vec![items[i].clone()],
            items.iter()
                 .enumerate()
                 .filter(|(idx, _x)| *idx != i)
                 .map(|(_idx, x)| x.clone())
                 .collect::<Vec<Thing>>()
        ));

        // Pick 2 items
        for j in i+1 .. items.len() {
            result.push((
                vec![items[i].clone(), items[j].clone()],
                items.iter()
                    .enumerate()
                    .filter(|(idx, _x)| *idx != i && *idx != j)
                    .map(|(_idx, x)| x.clone())
                    .collect::<Vec<Thing>>()
            ));
        }
    }
    result
}

// If microchip A and any generator, then have to have generator A
fn is_acceptable(things: &Vec<Thing>) -> bool {
    let any_generator = things.iter().any(|thing| matches!(thing, Thing::Generator(_)));
    /*
    if !any_generator {
        return true;
    }
    */
    ! any_generator
    || things.iter().all(|thing| {
        match thing {
            Thing::Generator(_) => true,
            Thing::Microchip(chip_type) => things.contains(&Thing::Generator(chip_type.clone())),
        }
    })
    /*
    for thing in things {
        if let Thing::Microchip(thing_type) = thing {
            if !things.contains(Things::Generator(thing_type)) {
                return false;
            }
        }
    }
    true
    */
}

// Rules:
// At least 1, at most 2 things in the elevator in each turn
// If microchip A and any generator, then have to have generator A
fn move_to_fourth(start_state: &State) -> u32 {
    // queue of (step counter, state), for BFS
    let mut q: VecDeque<(u32, State)> = VecDeque::new();
    let mut visited: HashSet<String> = HashSet::new();
    q.push_back((0, start_state.clone()));
    while !q.is_empty() {
        let (steps, state) = q.pop_front().unwrap();
        if State::is_final(&state) {
            return steps;
        }
        // Reject state if already visited
        let state_str = state.as_string();
        //print!("steps {} state {}\n", steps, state_str);
        if visited.contains(&state_str) {
            continue;
        }
        visited.insert(state_str);
        // Pick any 1 or 2 items from the current floor.
        // There always has to be at least one item on the current
        // floor, otherwise we wouldn't be able to use the lift.
        let current_items = &state.floors[state.current];
        assert!(current_items.len() >= 1);
        for (choice, remain) in pick_one_or_two(current_items) {
            if !is_acceptable(&choice) || !is_acceptable(&remain) {
                continue;
            }
            // Move 1 floor up or down
            for change in [1, -1] {
                let new_floor = (state.current as i32) + change;
                // Verify that the state would still be valid
                if new_floor < 0 || new_floor > 3 {
                    continue;
                }
                let mut floor_after_move = state.floors[new_floor as usize].clone();
                let mut moved_items = choice.clone();
                floor_after_move.append(&mut moved_items);
                if is_acceptable(&floor_after_move) {
                    // Push new state to the back of the queue
                    let mut new_state = state.clone();
                    new_state.current = new_floor as usize;
                    new_state.floors[state.current] = remain.clone();
                    new_state.floors[new_floor as usize] = floor_after_move;
                    q.push_back((steps + 1, new_state));
                }
            }
        }
    }

    panic!("Solution not found");
}

fn main() {
    let mut file = File::open("input11.txt").unwrap();
    let mut text = String::new();
    file.read_to_string(&mut text).unwrap();
    let state = make_state(parse_input(&text));
    let num_steps = move_to_fourth(&state);
    println!("Part 1, steps={}", num_steps);
}

#[cfg(test)]
mod tests {
    use super::*;

    // 1: HM LM
    // 2: HG
    // 3: LG
    // 4:
    static TEST_INPUT: &str =
        "The first floor contains a hydrogen-compatible microchip and a lithium-compatible microchip.\n\
        The second floor contains a hydrogen generator.\n\
        The third floor contains a lithium generator.\n\
        The fourth floor contains nothing relevant.\n";
    
    #[test]
    fn test_parsing() {
        let floors = parse_input(TEST_INPUT);
        assert_eq!(floors.len(), 4);
        assert!(floors.contains_key(&1));
        assert!(floors.contains_key(&2));
        assert!(floors.contains_key(&3));
        assert!(floors.contains_key(&4));
        assert_eq!(floors[&1], vec![Thing::Microchip("hydrogen".to_string()), Thing::Microchip("lithium".to_string())]);
        assert_eq!(floors[&2], vec![Thing::Generator("hydrogen".to_string())]);
        assert_eq!(floors[&3], vec![Thing::Generator("lithium".to_string())]);
        assert_eq!(floors[&4], vec![]);
    }
    
    #[test]
    fn test_state_as_string() {
        let state = State {
            floors: [
                vec![Thing::Generator("F".to_string()), Thing::Microchip("F".to_string())],
                vec![Thing::Microchip("A".to_string())],
                vec![Thing::Generator("A".to_string())],
                vec![]
            ],
            current: 0,
        };
        let state_string = State::as_string(&state);
        assert_eq!(state_string, "GF,MF::MA::GA:: 0".to_string());
    }
    
    #[test]
    fn test_state_as_string_sorting() {
        let state = State {
            floors: [
                vec![Thing::Microchip("A".to_string()), Thing::Generator("F".to_string()), Thing::Generator("A".to_string())],
                vec![],
                vec![],
                vec![],
            ],
            current: 0,
        };
        let state_string = State::as_string(&state);
        assert_eq!(state_string, "GA,GF,MA:::::: 0".to_string());
    }

    #[test]
    fn test_final_state() {
        let state = State {
            floors: [
                vec![Thing::Microchip("A".to_string())],
                vec![],
                vec![],
                vec![Thing::Generator("A".to_string())],
            ],
            current: 3,
        };
        assert!(!State::is_final(&state));

        let final_state = State {
            floors: [
                vec![],
                vec![],
                vec![],
                vec![Thing::Microchip("A".to_string()),Thing::Generator("A".to_string())],
            ],
            current: 3,
        };
        assert!(State::is_final(&final_state));
    }

    #[test]
    fn test_pick_one_or_two() {
        let m_a = Thing::Microchip("A".to_string());
        let m_b = Thing::Microchip("B".to_string());
        let g_a = Thing::Generator("A".to_string());

        let items = vec![
            m_a.clone(),
            m_b.clone(),
            g_a.clone(),
        ];
        let choices = pick_one_or_two(&items);
        assert_eq!(choices.len(), 6);  // 3 single choices and 3 double choices
        assert_eq!(choices[0], (vec![m_a.clone()], vec![m_b.clone(), g_a.clone()]));
        assert_eq!(choices[1], (vec![m_a.clone(), m_b.clone()], vec![g_a.clone()]));
        assert_eq!(choices[2], (vec![m_a.clone(), g_a.clone()], vec![m_b.clone()]));
        assert_eq!(choices[3], (vec![m_b.clone()], vec![m_a.clone(), g_a.clone()]));
        assert_eq!(choices[4], (vec![m_b.clone(), g_a.clone()], vec![m_a.clone()]));
        assert_eq!(choices[5], (vec![g_a.clone()], vec![m_a.clone(), m_b.clone()]));
    }

    #[test]
    fn test_is_acceptable() {
        let m_a = Thing::Microchip("A".to_string());
        let m_b = Thing::Microchip("B".to_string());
        let g_a = Thing::Generator("A".to_string());
        let g_b = Thing::Generator("B".to_string());

        // acceptable combinations
        assert!(is_acceptable(&vec![]));
        assert!(is_acceptable(&vec![m_a.clone()]));
        assert!(is_acceptable(&vec![g_a.clone()]));
        assert!(is_acceptable(&vec![m_a.clone(), g_a.clone()]));
        assert!(is_acceptable(&vec![m_a.clone(), m_b.clone()]));
        assert!(is_acceptable(&vec![g_a.clone(), g_b.clone()]));
        assert!(is_acceptable(&vec![m_a.clone(), g_a.clone(), g_b.clone()]));
        // not acceptable combinations
        assert!(!is_acceptable(&vec![m_a.clone(), g_a.clone(), m_b.clone()]));
        assert!(!is_acceptable(&vec![m_a.clone(), g_b.clone()]));
    }

    #[test]
    fn test_move_to_fourth() {
        let state = make_state(parse_input(TEST_INPUT));
        let step_count = move_to_fourth(&state);
        assert_eq!(step_count, 11);
    }

}

File: 2016/src/11.py
#!/usr/bin/env python3.8
import re
from collections import deque
import heapq

EXAMPLE = """The first floor contains a hydrogen-compatible microchip and a lithium-compatible microchip.
The second floor contains a hydrogen generator.
The third floor contains a lithium generator.
The fourth floor contains nothing relevant."""

EXPECTED = 11

MICROCHIP = 0
GENERATOR = 1

MATERIALS = []

class Thing:
    # Represent material as an index (integer) to the MATERIALS list to be able
    # to deal with numbers instead of strings
    def __init__(self, material, kind):
        dash = material.find('-')
        if dash > -1:
            material = material[:dash]

        global MATERIALS
        if material not in MATERIALS:
            MATERIALS.append(material)
        self.material = MATERIALS.index(material)
        self.kind = MICROCHIP if kind == 'microchip' else GENERATOR

    def __repr__(self):
        return f"{'M' if self.kind == MICROCHIP else 'G'}({MATERIALS[self.material]})"


def set_of_ones(val):
    c = 0
    ones = []
    while val > 0:
        if val & 1:
            ones.append(c)
        val //= 2
        c += 1
    return ones


class Floor:
    # Represent each floor as a bitmap of microchips and of generators
    # More efficient than keeping a list of things, assuming the number
    # of different materials is quite low.
    def __init__(self, chips, generators, num_things):
        self.chips = chips
        self.generators = generators
        self.num_things = num_things

    @classmethod
    def from_things(cls, things):
        chips = 0
        generators = 0
        for thing in things:
            if thing.kind == MICROCHIP:
                chips |= 2 ** thing.material
            else:
                generators |= 2 ** thing.material
        return Floor(chips, generators, len(things))

    def __repr__(self):
        def get_materials(bitmap):
            return [MATERIALS[c] for c in set_of_ones(bitmap)]
        things = [f"M({mat})" for mat in get_materials(self.chips)] + \
            [f"G({mat})" for mat in get_materials(self.generators)]
        return ', '.join(things)

    def empty(self):
        return self.chips == 0 and self.generators == 0

    def get_one_or_two_things(self):
        # yield (picked_microchips, picked_generators, new_floor_after_picking)
        # only if new_floor_after_picking is valid
        chips = set_of_ones(self.chips)
        gens = set_of_ones(self.generators)

        def all_options():
            # pick one or two chips
            for i in range(len(chips)):
                first = 2 ** chips[i]
                yield (first, 0, 1)
                for j in range(i+1, len(chips)):
                    second = 2 ** chips[j]
                    yield (first | second, 0, 2)

            # pick one or two generators
            for i in range(len(gens)):
                first = 2 ** gens[i]
                yield (0, first, 1)
                for j in range(i+1, len(gens)):
                    second = 2 ** gens[j]
                    yield (0, first | second, 2)

            # pick one chip and one generator
            for i in chips:
                for j in gens:
                    yield (2**i, 2**j, 2)

        chip_and_its_gen = False
        for opt_chip, opt_gen, num_things in all_options():

            if opt_chip !=0 and opt_gen != 0 and opt_chip & opt_gen == opt_chip:
                # [this optimization reduced states from 2.7M to 2.15M in part 2]
                # we're picking a chip and its generator
                # If there are multiple different pairs of chips and their own
                # generators on the same floor, we don't need to try all the options,
                # just one of them. The solution will be equivalent, no matter what
                # material and chip do we pick
                if chip_and_its_gen:
                    continue
                chip_and_its_gen = True

            new_floor = self.withoutThings(opt_chip, opt_gen, num_things)
            if new_floor.is_valid():
                yield (opt_chip, opt_gen, new_floor, num_things)

    def withExtraThings(self, add_chips, add_generators, num_things):
        return Floor(self.chips | add_chips, self.generators | add_generators, self.num_things + num_things)
    
    def withoutThings(self, remove_chips, remove_generators, num_things):
        return Floor(self.chips & ~remove_chips, self.generators & ~remove_generators, self.num_things - num_things)

    def is_valid(self):
        # floor is valid if 
        # a) there are no generators
        if self.generators == 0:
            return True
        # or b) each chip present has its generator also present
        return self.chips & self.generators == self.chips

    def __lt__(self, other):
        # Make two Floor instances comparable to make the heapq happy, but otherwise
        # there isn't any particular reason for this ordering.
        return (self.chips, self.generators) < (other.chips, other.generators)


class HashKey:
    def __init__(self, current, floors):
        values = [current]
        values += [fl.chips for fl in floors]
        values += [fl.generators for fl in floors]
        self.values = tuple(values)

    def __hash__(self):
        return hash(self.values)
    
    def __eq__(self, other):
        return other and self.values == other.values
    
    def __ne__(self, other):
        return not (self == other)


def parse_input(lines):
    assert len(lines) == 4
    pat = re.compile(r'([^ ]+) (microchip|generator)')
    floors = []
    for ln in lines:
        things = pat.findall(ln)
        floors.append(Floor.from_things([Thing(mat, kind) for mat, kind in things]))
    return floors


def end_state(current, floors):
    return current == 3 and floors[0].empty() and floors[1].empty() and floors[2].empty()


def heuristic(current, floors):
    # Lower bound (underestimated) number of steps necessary to bring all items from the given
    # configuration to the fourth floor
    lowest_nonempty_floor = 3
    cost = 0
    for i in range(3):
        items = floors[i].num_things
        if items > 0:
            if i < lowest_nonempty_floor:
                lowest_nonempty_floor = i
                if current > i:
                    # we will still have to go down from current to 'i' to pick up what's there
                    cost += current - i
            if i != current:
                # [reduced number of states from 2.15M to 1.55M in part 2]
                # When we go to this floor, we will be carrying at least one additional item with us
                items += 1

            # To pick up all items, we need to visit the floor at least items -1 times
            # After each visit, when there are any items left, we still need to return to the
            # floor, which is one extra step for each additional item over 2
            steps_up = max(1, items - 1)
            steps_down = max(0, items - 2)
            # All the way to the fourth floor
            cost += (3 - i) * (steps_up + steps_down)
    return cost


# BFS, ~158000 states visited in part 1
def min_steps_to_4_bfs(floors):
    q = deque()
    q.append((0, 0, floors))
    visited = set()
    while q:
        steps, current, floors = q.popleft()
        if end_state(current, floors):
            print(f"States visited: {len(visited)}")
            return steps
        key = HashKey(current, floors)
        if key in visited:
            continue
        visited.add(key)
        # try picking one or two items from the current floor
        for pick_chip, pick_gener, new_source_floor, num_things in floors[current].get_one_or_two_things():
            # try to go 1 floor up or down
            for change in [1, -1]:
                new_floor_nr = current + change
                if new_floor_nr < 0 or new_floor_nr >= 4:
                    # out of bounds
                    continue
                new_dest_floor = floors[new_floor_nr].withExtraThings(pick_chip, pick_gener, num_things)
                if not new_dest_floor.is_valid():
                    # movin stuff to the new floor would make it imbalanced
                    continue
                # move is valid
                new_floors = floors[:]
                new_floors[current] = new_source_floor
                new_floors[new_floor_nr] = new_dest_floor
                q.append((steps + 1, new_floor_nr, new_floors))
    assert False, "solution not found"

# A*, ~31k states visited in part 1, 3.3M states visited in part 2
def min_steps_to_4(floors):
    q = []
    print(f"Initial state cost from heuristic: {heuristic(0, floors)}")
    heapq.heappush(q, (heuristic(0, floors), 0, 0, floors))
    visited = set()
    while q:
        _, steps, current, floors = heapq.heappop(q)
        if end_state(current, floors):
            print(f"States visited: {len(visited)}")
            return steps
        key = HashKey(current, floors)
        if key in visited:
            continue
        visited.add(key)
        # try picking one or two items from the current floor
        for pick_chip, pick_gener, new_source_floor, num_things in floors[current].get_one_or_two_things():
            # try to go 1 floor up or down
            for change in [1, -1]:

                # [with this optimization, down from 3.3M to 2.7M states in part 2]:
                # does it ever make sense to just carry down 2 microchips?
                if change == -1 and pick_gener == 0 and num_things == 2:
                    continue

                new_floor_nr = current + change
                if new_floor_nr < 0 or new_floor_nr >= 4:
                    # out of bounds
                    continue
                new_dest_floor = floors[new_floor_nr].withExtraThings(pick_chip, pick_gener, num_things)
                if not new_dest_floor.is_valid():
                    # movin stuff to the new floor would make it imbalanced
                    continue
                # move is valid
                new_floors = floors[:]
                new_floors[current] = new_source_floor
                new_floors[new_floor_nr] = new_dest_floor
                heapq.heappush(q, (steps + 1 + heuristic(new_floor_nr, new_floors), steps + 1, new_floor_nr, new_floors))
    assert False, "solution not found"


def solve(floors):
    for i, floor in enumerate(floors):
        print(f"{i+1}: {floor}")

    return min_steps_to_4(floors)


if __name__ == "__main__":
    # Test for part 1
    test_floors = parse_input(EXAMPLE.split('\n'))
    test_value = solve(test_floors)
    print(f"Test result: {test_value}")
    assert test_value == EXPECTED, f"Test expected {EXPECTED}, got {test_value}"
    print()

    # clear global state (sic)
    MATERIALS = []

    # Actual part 1
    with open('../input11.txt', 'rt') as f:
        actual_input = f.readlines()
    floors_pt1 = parse_input(actual_input)
    part1 = solve(floors_pt1)
    print(f"Part 1: {part1}")
    print()

    # Actual part 2
    floors_pt2 = parse_input(actual_input)
    extra_things = [
        Thing('elerium', 'generator'),
        Thing('elerium-compatible', 'microchip'),
        Thing('dilithium', 'generator'),
        Thing('dilithium-compatible', 'microchip')
    ]
    dummy_floor = Floor.from_things(extra_things)
    floors_pt2[0] = floors_pt2[0].withExtraThings(dummy_floor.chips, dummy_floor.generators, dummy_floor.num_things)
    part2 = solve(floors_pt2)
    print(f"Part 2: {part2}")
